src/librustc_mir/transform/const_prop.rs

```diff
@@ -636,18 +643,40 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     ScalarMaybeUndef::Scalar(one),
                     ScalarMaybeUndef::Scalar(two)
                 ) => {
-                    let ty = &value.layout.ty.kind;
+                    // Value represented as a pair, but we still want to generate zsts for rest of
+                    // the fields.
+                    let layout = &value.layout;
+                    debug!("replace_with_const: value layout={:#?}", layout);
+                    debug!("replace_with_const: value field count={}", layout.fields.count());
+                    let ty = &layout.ty.kind;
                     if let ty::Tuple(substs) = ty {
+                        // Values from the pair will be popped from this vec when we see a non-zst
+                        // type in the layout
+                        let mut pair_values = vec![two, one];
```

Comment:
Instead of allocating a new vector, we could do something like `let mut pair_values = &[one, two].iter()` and then around line 666 below, you'd just do `pair_values.next().unwrap()`.